odoo/addons/base/ir/ir_translation.py

```diff
@@ -121,18 +121,27 @@ def finish(self):
         # referencing non-existent data.
         cr.execute("DELETE FROM %s WHERE res_id IS NULL AND module IS NOT NULL" % self._table)
 
+        # detect the xml_translate fields, where the src must be the same
+        env = api.Environment(cr, SUPERUSER_ID, {})
+        src_relevant_fields = []
+        for model in env:
+            for field_name, field in env[model]._fields.items():
+                if hasattr(field, 'translate') and callable(field.translate):
+                    src_relevant_fields.append("%s,%s" % (model, field_name))
+
         find_expr = """
                 irt.lang = ti.lang
             AND irt.type = ti.type
             AND irt.name = ti.name
             AND (
-                    (ti.type = 'model' AND ti.res_id = irt.res_id AND irt.src = ti.src)
+                    (ti.type = 'model' AND ti.res_id = irt.res_id AND ti.name IN %s AND irt.src = ti.src)
+                 OR (ti.type = 'model' AND ti.res_id = irt.res_id AND ti.name NOT IN %s)
                  OR (ti.type = 'view' AND (irt.res_id IS NULL OR ti.res_id = irt.res_id) AND irt.src = ti.src)
                  OR (ti.type = 'field')
                  OR (ti.type = 'help')
                  OR (ti.type NOT IN ('model', 'view', 'field', 'help') AND irt.src = ti.src)
             )
-        """
+        """ % (tuple(src_relevant_fields), tuple(src_relevant_fields))
```

Comment:
Technically I'm not sure we can consider field names and model names SQL-safe. Granted, modifying them requires admin access and they should only contain alphanumeric characters. But there is still a chance admins might screw up their databases this way and crash the query.
It's uglier but we should consider passing the list as parameters to `execute`.